platforms/core-configuration/model-core/src/main/java/org/gradle/internal/extensibility/DefaultExtraPropertiesExtension.java

```diff
@@ -79,7 +99,16 @@ public void setProperty(String name, @Nullable Object newValue) {
 
     @Override
     public Map<String, Object> getProperties() {
-        return new HashMap<>(storage);
+        // TODO:configuration-cache use a tracking map here
+        HashMap<String, Object> properties = new HashMap<>(storage);
+        for (Map.Entry<String, Object> entry : gradleProperties.entrySet()) {
+            if (!properties.containsKey(entry.getKey())) {
```

Comment:
❓ 


Why not just 

`properties.putIfAbsent(entry.getKey(), entry.getValue());`

Reply:
The `if` is also guarding the (commented out) `onGradlePropertyLookup`.